test/object_store/conftest.py

```diff
@@ -108,3 +109,15 @@ async def s3_server(pytestconfig, tmpdir):
         yield server
     finally:
         await server.stop()
+
+@pytest.fixture(scope="function")
+async def s3_client(s3_server):
```

Comment:
i think `s3_client()` should be a free helper function instead of a fixture. because, well, it's not a a fixture..

Reply:
How do I tell fixtures from non-fixtures?

Reply:
i think fixtures are more like running environment or a dataset, which is relatively fixed. i'd defer this to something like https://docs.pytest.org/en/stable/explanation/fixtures.html .

Reply:
> If s3_server is a fixture, and s3 is something built from s3_server, then it does make sense for it to be a fixture too. What is strange, however, is why both s3_server and s3_client are **function**-scoped fixtures. Is this deliberate? These are inefficient (a new one is created separately for each test function) and should be used sparingly if you want tests to be fast. It also raises @tchaikov's question of why it's even a "fixture" and not simply a contextmanager which the test calls.

it's not that inefficient when we drive the tests using `test.py` , as `s3_server` fixture only returns a struct to the test,  and the struct just carries the address and port number for connecting to the minio server.

regarding why its function-scoped fixture, i'd prefer using function-scoped fixture when possible because i wanted to make each test isolated so they don't have look at the leftovers of other tests if any. probably this is controversial, i still recall the debate that Pavel and i had. so i don't insist.

> 
> One of the nice features of function-scoped (if you must?) fixtures over contextmanager is reducing indentation of the test function - so the test doesn't need to begin with a "with" scope. I agree it's not a huge advantage - non-function scoped fixtures have more important advanteges.